scheduler/scheduler-server/src/test/java/org/ow2/proactive/scheduler/synchronization/AOSynchronizationTest.java

```diff
@@ -0,0 +1,501 @@
+            // expected
+        }
+
+        Runnable decrementRunnable = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    Thread.sleep(1000);
+                    synchronization.compute(CHANNEL1, "a", BIFUNCTION_DECREMENT_ONE);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+
+            }
+        };
+
+        // this is a producer/consumer pattern
+        // the producer will decrement the counter, consumer will wait until the counter reach zero
+        executor.submit(decrementRunnable);
+        Assert.assertTrue(synchronization.waitUntil(CHANNEL1, "a", PREDICATE_EQ_ZERO));
+        Assert.assertEquals(0, (int) synchronization.get(CHANNEL1, "a"));
+    }
+
+    @Test
+    public void testWaitUntilWithTimeout() throws IOException, InvalidChannelException, CompilationException {
+        initChannel();
+
+        Runnable decrementRunnable = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    Thread.sleep(1000);
+                    synchronization.compute(CHANNEL1, "a", BIFUNCTION_DECREMENT_ONE);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+
+            }
+        };
+
+        // this is a producer/consumer pattern
+        // the producer will decrement the counter, consumer will wait until the counter reach zero
+        // a timeout exception should be thrown by the wait method
+        executor.submit(decrementRunnable);
+        try {
+            synchronization.waitUntil(CHANNEL1, "a", PREDICATE_EQ_ZERO, 500);
+            fail("TimeoutException should be thrown");
+        } catch (TimeoutException e) {
+            // expected
+        }
+    }
+
+    @Test
+    public void testWaitUntilThen() throws IOException, InvalidChannelException, CompilationException {
+        initChannel();
+        try {
+            synchronization.waitUntilThen(CHANNEL1,
+                                          "a",
+                                          PREDICATE_WITH_COMPILATION_ERROR,
+                                          BIFUNCTION_WITH_COMPILATION_ERROR);
+            fail("CompilationException should be thrown");
+        } catch (CompilationException e) {
+            // expected
+        }
+
+        try {
+            synchronization.waitUntilThen(CHANNEL1,
+                                          "a",
+                                          PREDICATE_WITH_RUNTIME_ERROR,
+                                          BIFUNCTION_WITH_COMPILATION_ERROR);
+            fail("ClosureEvaluationException should be thrown");
+        } catch (ClosureEvaluationException e) {
+            // expected
+        }
+
+        Runnable decrementRunnable = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    Thread.sleep(1000);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)